### PR TITLE
tcp_check to reconnect in case of connection lost

### PR DESCRIPTION
**Pull request by afeno**
**tcp_check is trying to send an internal message every 10 seconds to test
the connection even if there is no activity while the socket is still alive.**

These are the changes:
1) Modify the `is_sensor`and `set_child_value` to allow internal messages (id=255) to be sent
2) After `connect()`, start checking the connection every 10 seconds sending a HEARTBEAT message to the gateway.
_Note: If the sent() is not successful then the socket will be closed and the main thread will try to create a new connection every 10 seconds._
3) The disconnect() will stop the tcp_check because at this moment we know that the socket is closed.

### CODE REVIEW

File: mysensors/mysensors.py [hunk 1]
```diff
@@ -552,6 +558,15 @@ def _check_socket(self, sock=None, timeout=None):
             raise OSError
         return available_socks
 
+    def tcp_check(self):
+        if not self._stop_tcp_check.is_set():
+            t = threading.Timer(10.0, self.tcp_check)    #send heartbeat every 10 seconds until socket is disconnected
```
Comment (MartinHjelmare) [inline mysensors/mysensors.py]: Why do you start a new thread here? You just need to check time and compare that to the last time the tcp check was done. Save the time of the check in an instance variable when the check is done. If time is not enough just return.  Call the method from the run loop.

File: mysensors/mysensors.py [hunk 2]
```diff
@@ -552,6 +558,15 @@ def _check_socket(self, sock=None, timeout=None):
...
+            t.setDaemon(True)
+            t.start()
+            _LOGGER.info('Sending I_HEARTBEAT')
+            #self.set_child_internal(0, 255, 2, 0)
+            self.set_child_value(0, 255, 18, 0)
```
Comment (MartinHjelmare) [inline mysensors/mysensors.py]: You should not use this method. Create a Message and use the copy method to set it to internal type etc. Then add it to the FIFO queue with the fill_queue method, while calling the encode method. See https://github.com/afeno/pymysensors/blob/81d11f1f0d750c98819811031fdcbaff08531033/mysensors/mysensors.py#L322-L327.

So then you don't need to change set_child_value method or is_sensor method.

I would prefer if you can use internal message I_VERSION which is supported by mysensors 1.4 and higher. Please test that, if you haven't done it already. The heartbeat messages are new in version 2.0.

File: main.py [hunk 3]
```diff
@@ -7,11 +7,11 @@ def event(update_type, nid):
...
 
 # To create a serial gateway.
-GATEWAY = mysensors.SerialGateway(
-    '/dev/ttyACM0', event, protocol_version='2.0')
+#GATEWAY = mysensors.SerialGateway(
+#    '/dev/ttyACM0', event, protocol_version='2.0')
 
 # To create a TCP gateway.
-# GATEWAY = mysensors.TCPGateway('127.0.0.1', event, True)
+GATEWAY = mysensors.TCPGateway('192.168.1.84', event, True)
```
Comment (MartinHjelmare) [inline main.py]: Please don't change this file. It should only serve as an example. You can change it while testing but don't commit the change.

**Comment by afeno** [comment]
Thank you Martin for all the comments. All of them make sense. I will work on that and come back to you later with a new proposal.
Should I do something with this PR on my side? My intention is to re-work this and create a new PR. Is that correct?

**Comment by MartinHjelmare** [comment]
You don't need to open a new PR. You can keep committing to the same branch and then when you feel ready for a new review, just push again and github will update this PR.

Opening a new PR is suitable if you want to make a totally different approach. You can do as you like in this case.

**Comment by MartinHjelmare** [comment]
If you do want to start over and make a new PR, I advice you to create a new branch of your own that is based on the current upstream dev branch. When you work on a new feature or fix you should always do so in a new separate branch that you have created. It will make your life a lot easier when using git.

**Comment by afeno** [comment]
This is a new version taking into consideration the feedback and recommendations received. Basically, it ask for the `I_VERSION` every `reconnect_timeout` seconds using the `time.time()` function.

**Comment by afeno** [comment]
Ok. Got it. But... What is the intention of the second timer to disconnect the socket? The Send() will disconnect the socket anyway if the connection was lost. I think that I'm missing something...
Thanks for the code. It is almost done  :)

**Comment by afeno** [comment]
First of all, I apologise for the terminology that I will use in my explanation below:
The _check_socket method will only detect the disconnection when the disconnection happened at the pymysensors side (very rare I think).
If the gateway for example is suddenly turned off, pymysensors will not detect the disconnection until it tries to send something to it.
Imagine a gateway with only sensors attached to it (temperature sensors for example) that only send data to pymysensors. If the gateway goes down for a couple of minutes, the pymysensors will never detect it because it is just waiting to receive data. Only when pymysensors tries to send something it will realise that the connection was lost and then close the socket. (The socket will still be alive until something is sent). After the socket is closed then it will try to connect again.
This is my understanding about how it works and because of that reason I thought about sending something (Version request) every x seconds to the gateway.

**Comment by MartinHjelmare** [comment]
Sounds reasonable. Then skip the extra timer attribute etc, and you also don't really need to handle the response to the internal message. Make sure you test the solution live, when you are happy.

**Comment by afeno** [comment]
This version is pretty clean exactly as you posted above.
I tested it with one gateway and it is working fine on my side

`2016-10-25 16:32:15,017 - mysensors.mysensors - DEBUG - Sending 0;255;3;0;2;
2016-10-25 16:32:15,669 - mysensors.mysensors - DEBUG - Received 0;255;3;0;2;2.0.0
...
... Gateway turned off
...
2016-10-25 16:32:25,027 - mysensors.mysensors - DEBUG - Sending 0;255;3;0;2;
2016-10-25 16:32:33,117 - mysensors.mysensors - ERROR - Receive from server failed.
2016-10-25 16:32:33,117 - mysensors.mysensors - INFO - Closing socket at ('192.168.1.84', 5003).
2016-10-25 16:32:33,117 - mysensors.mysensors - INFO - Socket closed at ('192.168.1.84', 5003).
2016-10-25 16:32:33,117 - mysensors.mysensors - INFO - Trying to connect to ('192.168.1.84', 5003)
2016-10-25 16:32:43,155 - mysensors.mysensors - ERROR - Failed to connect to socket at ('192.168.1.84', 5003).
2016-10-25 16:32:43,155 - mysensors.mysensors - INFO - Waiting 10 secs before trying to connect again.
2016-10-25 16:32:53,163 - mysensors.mysensors - INFO - Trying to connect to ('192.168.1.84', 5003)
2016-10-25 16:33:03,166 - mysensors.mysensors - ERROR - Failed to connect to socket at ('192.168.1.84', 5003).
2016-10-25 16:33:03,166 - mysensors.mysensors - INFO - Waiting 10 secs before trying to connect again.
2016-10-25 16:33:13,148 - mysensors.mysensors - INFO - Trying to connect to ('192.168.1.84', 5003)
2016-10-25 16:33:23,187 - mysensors.mysensors - ERROR - Failed to connect to socket at ('192.168.1.84', 5003).
2016-10-25 16:33:23,187 - mysensors.mysensors - INFO - Waiting 10 secs before trying to connect again.
...
...Gateway turned on
...
2016-10-25 17:11:14,998 - mysensors.mysensors - INFO - Trying to connect to ('192.168.1.84', 5003)
2016-10-25 17:11:15,326 - mysensors.mysensors - INFO - Connected to ('192.168.1.84', 5003)
2016-10-25 17:11:15,373 - mysensors.mysensors - DEBUG - Sending 0;255;3;0;2;
2016-10-25 17:11:15,435 - mysensors.mysensors - DEBUG - Received 0;255;3;0;14;Gateway startup complete.
2016-10-25 17:11:15,653 - mysensors.mysensors - DEBUG - Received 0;255;3;0;11;Cuartito
sensor_update 0
2016-10-25 17:11:15,887 - mysensors.mysensors - DEBUG - Received 0;255;3;0;12;1.0`

**Comment by MartinHjelmare** [comment]
Why does it take seven seconds between sending a request for version, and the error to receive? Also, shouldn't the send operation fail, as you said before? Now it's the receive operation that fails.

**Comment by afeno** [comment]
Ok. I'm learning a bit more about sockets :) and this is my understanding:
When we call to the socket.send() the data is delivered to the TCP stack and it's up to the TCP stack to deliver it when it can. Because of this, the socket.send() will not return an error intermediately when the gateway is dead.
However, it seems that socket.recv() in combination with select.select can be used to detect a broken connection (errno 10054). Note that the detection/error will only occur if we try to send something first so the _check_connection() is still necessary. 
What I don't know yet is why we have 7 seconds between the message sent and the 10054 recv error. Maybe this is the TCP timeout? 

`2016-10-27 00:35:20,685 - mysensors.mysensors - DEBUG - Sending 0;255;3;0;2;
2016-10-27 00:35:20,745 - mysensors.mysensors - DEBUG - Received 0;255;3;0;2;2.0.0
2016-10-27 00:35:30,687 - mysensors.mysensors - DEBUG - Sending 0;255;3;0;2;
2016-10-27 00:35:37,630 - mysensors.mysensors - ERROR - Receive from server failed. Errno 10054: An existing connection was forcibly closed by the remote host
2016-10-27 00:35:37,631 - mysensors.mysensors - INFO - Closing socket at ('192.168.1.53', 5003).
2016-10-27 00:35:37,631 - mysensors.mysensors - INFO - Socket closed at ('192.168.1.53', 5003).
2016-10-27 00:35:37,631 - mysensors.mysensors - INFO - Trying to connect to ('192.168.1.53', 5003)`

**Comment by MartinHjelmare** [comment]
Ok, well if this PR solves the issue compared to previous code, I think it's good for merging. Just give me the go ahead, and I'll merge.

**Comment by gtortone** [comment]
Hi, I would like to test this feature, how can I install it from git on my homeassistant machine ?

Thanks.

**Comment by MartinHjelmare** [comment]
@gtortone 

You should be able to by first uninstalling the existing pymysensors package:
```bash
pip uninstall pymysensors
```
Make sure you also delete any `pymysensors` or `mysensors` related directories in `~/.homeassistant/deps/`
Then install the PR branch:
```
pip3 install https://github.com/afeno/pymysensors/archive/1b784bde273ca3d14f3a03d2606c924365d22c98.zip#pymysensors==0.8.dev0
```
Then start `hass` with `--skip-pip` option:
```
hass --skip-pip
```

**Comment by gtortone** [comment]
@afeno @MartinHjelmare 

I added some socket handling on shutdown in order to avoid "broken pipe" error;

please check here:
https://github.com/afeno/pymysensors/pull/1

**Comment by afeno** [comment]
The solution is working but I'm not convinced about relaying on the tcp stack (OS) to rise the error. The TCP timeout is really different depending on the OS. Soon or later the TCP layer (OS) will detect the connection problem but I don't know yet how long it will take.

**Comment by MartinHjelmare** [comment]
Well, there is still my suggestion of a second timer and checking for a reply from the arduino gateway. If no reply within time, then disconnect.

**Comment by afeno** [comment]
Yes. I will have a look again to the second timer alternative.

File: mysensors/mysensors.py [hunk 4]
```diff
@@ -114,7 +114,9 @@ def _handle_heartbeat(self, msg):
 
     def _handle_internal(self, msg):
         """Process an internal protocol message."""
-        if msg.sub_type == self.const.Internal.I_ID_REQUEST:
+        if msg.sub_type == self.const.Internal.I_VERSION:
```
Comment (MartinHjelmare) [inline mysensors/mysensors.py]: You have to overwrite this method in the child class TCPGateway and add the logic there. Make sure you also call this parent method via `super` and handle the return value.

**Comment by afeno** [comment]
Tested with HA 0.38.2 and everything is working fine.

File: mysensors/mysensors.py [hunk 5]
```diff
@@ -502,9 +501,33 @@ def __init__(self, host, event_callback=None,
...
+            child_id=255, type=self.const.MessageType.internal,
+            sub_type=self.const.Internal.I_VERSION)
+        self.fill_queue(msg.encode)
+        self.tcp_check_timer = time.time()
+
+    def _handle_internal(self, msg):
+        if msg.sub_type == self.const.Internal.I_VERSION:
+            self.tcp_disconnect_timer = time.time()
+        else:
+            super()._handle_internal(msg)
```
Comment (MartinHjelmare) [inline mysensors/mysensors.py]: You have to return the result of this call to the logic method which makes the call to the method here.
```py
return super()._handle_internal(msg)
```

**Comment by MartinHjelmare** [comment]
Great! Thanks @afeno !

@gtortone will unfortunately not be included among the contributors when I squash and merge this into dev, due to how github works. If those changes would have been submitted as a separate PR directly to pymysensors, the commit would have been counted. I'll make sure to mention you when I bump version in master. I hope this is ok.

**Comment by gtortone** [comment]
no problem !
thanks for your message !

**Pull request merged**